llama_cpp/llama.py

```diff
@@ -205,6 +206,44 @@ def __call__(
     ) -> bool:
         return any([stopping_criteria(input_ids, logits) for stopping_criteria in self])
 
+# Custom data that is accessible to the beam_search_callback() function.
+@dataclass
+class beam_search_callback_data:
+    ctx: llama_cpp.llama_context_p
+    response_tokens: List[int]
+
+# Used for debugging to view beam states
+def beam_view_to_string(ctx, beam_view):
+    string = f"p({beam_view.p}): "
+    for i in range(beam_view.n_tokens):
+        string += llama_cpp.llama_token_get_text(ctx, beam_view.tokens[i]).decode("utf-8")
+    return string
+
+# One requirement of the callback is that it MUST determine end-of-beam.
+def is_at_eob(ctx, tokens, n_tokens) :
+    return 0 < n_tokens and tokens[n_tokens-1] == llama_cpp.llama_token_eos(ctx);
+
+# beam_search_callback requires a global dictionary to pass data via their object id.
+beam_search_dictionary = {}
+
+# beam_search_callback() must flag beams when they reach end-of-sentence.
+# TODO: Use stop_sequences.
```

Comment:
Is this TODO the reason I'm seeing this in the debug output? (note the EOS and BOS)
```
beams[0] p(0.493342787027359): <0x0A></s><s><0x0A>#include▁<iostream><0x0A>#include▁<cmath.h><0x0A>#include▁<c
beams[1] p(0.5066572427749634): <0x0A></s><s><0x0A>#include▁<iostream><0x0A>#include▁<cmath.h><0x0A>#include▁<vector
```

Reply:
I'm not sure what you mean exactly by "Note the EOS and BOS."

The TODO note relates to the `is_at_eob()` function above. Currently, EOB (end-of-beam) is determined by the character `llama_cpp.llama_token_eos(ctx)`. If EOB is to be generalized to user-defined EOB sequences, then this would be the function to add the logic to.

Reply:
What I mean is that `</s>` (EOS) is generated by the model, but the beam search keeps going (onto BOS, and then it starts making up something unrelated). I think this shouldn't happen.

Reply:
To be clear, this is not a custom stop sequence, this is just the regular EOS token (AFAIK), which is rendered this way in the output. You say EOB is determined by llama_token_eos, but that doesn't seem to work for me.

Reply:
One way to debug this is to modify the line above
```
 string += llama_cpp.llama_token_get_text(ctx, beam_view.tokens[i]).decode("utf-8")
```
to something that appends both the numeric token id `beam_view.tokens[i]` along with the decoded substring. If you're really encountering the `llama_token_eos()` token, last I checked, it should have token id `2`.

Reply:
Oops! Sorry, this is one is my fault. llama_token_eos takes model now, not ctx, and I missed that when I merged this PR into my local branch. Unfortunately, ctypes gives no indication of pointer type mismatches.